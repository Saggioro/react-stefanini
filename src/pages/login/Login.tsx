import React, { useState } from "react";

import { Container, LoginArea } from "./Login.styled";
import Input from "../../components/input/Input";
import Button from "../../components/button/Button";
import { Formik } from "formik";
import * as Yup from "yup";
import * as authActions from "../../store/actions/auth";
import User from "../../models/User";
import { useAppDispatch, useAppSelector } from "../../hooks/LocalReduxThunk";
import { useNavigate } from "react-router-dom";
interface ILogin {}

const Login: React.FunctionComponent = ({ ...props }: ILogin) => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const users = useAppSelector((state) => state.usersReducer.users);
  const [error, setError] = useState<string>();
  const initialState = { login: "", password: "" };
  const validation = Yup.object().shape({
    login: Yup.string().required(),
    password: Yup.string().required(),
  });
  return (
    <Container>
      <LoginArea>
        <Formik
          initialValues={initialState}
          onSubmit={async (values) => {
            try {
              const userLogging = new User(values.login, values.password);
              const findUser = users.find(
                (user) => user.name === userLogging.name
              );
              if (!findUser) {
                throw new Error("User not found!");
              }

              if (findUser.password !== userLogging.password) {
                throw new Error("Incorrect password!");
              }
              await dispatch(authActions.logIn(userLogging) as any);
              navigate('/home');
            } catch (e: any) {
              setError(e.message);
            }

          }}
          validationSchema={validation}
        >
          {({
            values,
            errors,
            touched,
            handleChange,
            handleSubmit,
          }) => (
            <form onSubmit={handleSubmit}>
              <Input
                value={values.login}
                onChange={handleChange}
                name="login"
                label="Login"
                error={touched.login ? errors.login : undefined}
              />
              <Input
                value={values.password}
                onChange={handleChange}
                name="password"
                label="Password"
                error={touched.password ? errors.password : undefined}
              />
              <Button>Sign in</Button>
              {error && (
                <p>{error}</p>
              )}
            </form>
          )}
        </Formik>
      </LoginArea>
    </Container>
  );
};

export default Login;
