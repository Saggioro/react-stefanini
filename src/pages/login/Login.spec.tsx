/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/no-debugging-utils */
import renderer from "react-test-renderer";
import { render, screen, fireEvent, act } from "@testing-library/react";
import Login from "./Login";
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}));

jest.mock('react-router-dom', () => ({
    useNavigate: jest.fn(),
  }));

jest.mock("../../hooks/LocalReduxThunk", () => {
  return {
    useAppSelector() {
      return [
        {
          name: "teste",
          password: "123",
        },
      ];
    },
    useAppDispatch: () => mockDispatch
  };
});

test("Input should render correctly", () => {
  const component = renderer.create(<Login />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Login form should login", async () => {
    const { container } = render(<Login />);
    const inputLogin = container.querySelector(`input[name="login"]`) as Element;
    const button = container.querySelector(`button[type="submit"]`) as Element;
    const inputPassword = container.querySelector(
      `input[name="password"]`
    ) as Element;
  
    await act(async () => {
      fireEvent.change(inputLogin, { target: { value: "teste" } });
      fireEvent.change(inputPassword, { target: { value: "123" } });
      fireEvent.click(button);
    });
  
    expect(screen.queryByText("User not found!")).not.toBeInTheDocument();
    expect(screen.queryByText("Incorrect password!")).not.toBeInTheDocument();
  });

test("Login form should throw error when name doesnt exists", async () => {
    const { container } = render(<Login />);
    const inputLogin = container.querySelector(`input[name="login"]`) as Element;
    const button = container.querySelector(`button[type="submit"]`) as Element;
    const inputPassword = container.querySelector(
      `input[name="password"]`
    ) as Element;
  
    await act(async () => {
      fireEvent.change(inputLogin, { target: { value: "teste3" } });
      fireEvent.change(inputPassword, { target: { value: "123" } });
      fireEvent.click(button);
    });
  
    expect(screen.getByText("User not found!")).toBeInTheDocument();
  });

  test("Login form should throw error when password doesnt match", async () => {
    const { container } = render(<Login />);
    const inputLogin = container.querySelector(`input[name="login"]`) as Element;
    const button = container.querySelector(`button[type="submit"]`) as Element;
    const inputPassword = container.querySelector(
      `input[name="password"]`
    ) as Element;
  
    await act(async () => {
      fireEvent.change(inputLogin, { target: { value: "teste" } });
      fireEvent.change(inputPassword, { target: { value: "321" } });
      fireEvent.click(button);
    });
  
    expect(screen.getByText("Incorrect password!")).toBeInTheDocument();
  });