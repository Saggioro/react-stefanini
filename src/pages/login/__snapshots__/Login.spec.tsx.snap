// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<div
  className="sc-bczRLJ jhqIRe"
>
  <div
    className="sc-gsnTZi iywiwN"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="sc-dkzDqf kSTTCZ"
      >
        <label
          className="sc-eCYdqJ gTqeOc"
        >
          Login
        </label>
        <input
          className="sc-hKMtZM drPumR"
          name="login"
          onChange={[Function]}
          type="text"
          value=""
        />
        <span
          className="sc-jSMfEi gelXOA"
        />
      </div>
      <div
        className="sc-dkzDqf kSTTCZ"
      >
        <label
          className="sc-eCYdqJ gTqeOc"
        >
          Password
        </label>
        <input
          className="sc-hKMtZM drPumR"
          name="password"
          onChange={[Function]}
          type="text"
          value=""
        />
        <span
          className="sc-jSMfEi gelXOA"
        />
      </div>
      <button
        className="sc-gKXOVf fjgzkO"
        type="submit"
      >
        Sign in
      </button>
    </form>
  </div>
</div>
`;
